File: src/irrgarten/Player.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;
import java.util.Arrays;
/**
 *
 * @author aulas
 */
public class Player {
    private static int MAX_WEAPONS = 2;
    private static int MAX_SHIELDS = 3;
    private static int INITIAL_HEALTH = 10;
    private static int HITS2LOSE = 3;
    
    private String name;
    private char number;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    private int consecutiveHits = 0;
    
    ArrayList<Weapon> listWeapon = new ArrayList<>();
    ArrayList<Shield> listShield = new ArrayList<>();
    private int listWeapon_uses = 0;
    private int listShield_uses = 0;
    private int firstElement = 1;
    
    public Player(char number,float intelligence,float strength){
        this.name = "Player#" + number;
        this.intelligence = intelligence;
        this.strength = strength;
    }
    
    public int getRow(){
        return row;
    }
    
    public int getCol(){
        return col;
    }
    
    public char getNumber(){
        return number;
    }
    
    public void resurrect(){
        for(int i=0;i<listWeapon_uses;i++){
            listWeapon.remove(firstElement);
        }
        
        for(int i=0;i<listShield_uses;i++){
            listShield.remove(firstElement);
        }
        
        listWeapon_uses = 0;
        listShield_uses = 0;
        health = INITIAL_HEALTH;
        consecutiveHits = 0;  
    }
    
    public void setPos(int row,int col){
        this.row = row;
        this.col = col;
    }
    
    public boolean dead(){
        boolean alive = true;
        if(health < 1){
            alive = false;
        }
        return alive;
    }
    
    public Directions move(Directions direction,Directions[] validMoves){
        int size = validMoves.length;
        boolean contained = Arrays.asList(validMoves).contains(direction);
        if(size > 0 && !contained){
            Directions firstElement = validMoves[0];
            return firstElement;
        }else{
            return direction;
        }
    }
    
    public float attack(){
        return strength + sumWeapons();
    }
    
    public boolean defend(float receivedAttack){
        return manageHit(receivedAttack);
    }
    
    public void receiveReward(){
        int wReward = Dice.weaponReward();
        int sReward = Dice.shieldsReward();
        
        for(int i = 1 ; i <= wReward;i++){
            Weapon wnew = newWeapon();
            receiveWeapon(wnew);
        }
        
        for(int i = 1 ; i <= sReward;i++){
            Shield snew = newShield();
            receiveShield(snew);
        }
        
        int extraHealth = Dice.healthReward();
        
        health += extraHealth;
    }
    @Override
    public String toString(){
        String devolver = "Name :" +name +" " +
               "Intelligence :" + intelligence + " " +
               "Strength : " + strength + " " +
               "Health :" + health + " " +
               "Position : [" + row + "," + col + "] \n" ;
        devolver += "Weapons:\n";
        for(int i=0;i<listWeapon.size();i++){
            devolver += listWeapon.get(i).toString() + "\n";
        }
        devolver += "Shields:\n";
        for(int i=0;i<listShield.size();i++){
            devolver += listShield.get(i).toString() + "\n";
        }
        
        
        return devolver;
    }
    
    private void receiveWeapon(Weapon w){
        for(int i=listWeapon.size() -1;i> 0;i--){
            Weapon wi = listWeapon.get(i);
            boolean discard = wi.discard();
            
            if(discard){
                listWeapon.remove(wi);
            }
            int size = listWeapon.size();
            if(size < MAX_WEAPONS){
                listWeapon.add(w);
            }
        }
    }
    
    private void receiveShield(Shield s){
        for(int i= listShield.size()-1; i>0;i--){
            Shield si = listShield.get(i);
            boolean discard = si.discard();
            
            if(discard){
                listShield.remove(si);
            }
            
        }
        int size = listShield.size();
        if(size < MAX_SHIELDS){
            listShield.add(s);
        }
    }
    
    private Weapon newWeapon(){
        Weapon armament = new Weapon(Dice.weaponPower(),Dice.usesLeft());
        return armament;
    }
    
    private Shield newShield(){
        Shield protector = new Shield(Dice.shieldPower(),Dice.usesLeft());
        return protector;
    }
    
    
    private float sumWeapons(){
        float sum = 0;
        
        for(int i=0;i<listWeapon_uses ;i++){
            sum += listWeapon.get(i).attack();
        }
        
        return sum;
    }
    
    private float sumShields(){
        float sum = 0;
        
        for(int i=0;i<listShield_uses ;i++){
            sum += listShield.get(i).protect();
        }
        
        return sum;
    }
    
    private float defensiveEnergy(){
        return intelligence + this.sumShields();
    }
    
    private boolean manageHit(float receivedAttack){
        float defense = defensiveEnergy();
        if(defense < receivedAttack){
            gotWounded();
            incConsecutiveHits();
        }else{
            resetHits();
        }
        boolean lose;
        if(consecutiveHits == HITS2LOSE || dead()){
            resetHits();
            lose = true;
        }else{
            lose = false;
        }
        return lose;
    }
    
    private void resetHits(){
        this.consecutiveHits = 0;
    }
    
    private void gotWounded(){
        this.health--;
    }
    
    private void  incConsecutiveHits(){
        this.consecutiveHits++;
    }
}

File: RubyIrrgarten/Irrgarten1/Player.rb
require_relative 'Dice'
require_relative 'Weapon'
require_relative 'Shield'
require_relative 'Directions'

module Irrgarten
  class Player
    @@MAX_WEAPONS = 2
    @@MAX_SHIELDS = 3
    @@INITIAL_HEALTH = 10
    @@HITS2LOSE = 3

    def initialize(number,intelligence,strength)
      @weapons = Array.new
      @shields = Array.new
      @name = "Player##{number}"
      @number = number
      @intelligence = intelligence
      @strength = strength
      @health = @@INITIAL_HEALTH
      @row = 0
      @col = 0
      @consecutive_hits = 0
    end

    def resurrect
      @weapons.clear
      @shields.clear
      @health = @@INITIAL_HEALTH
      @consecutive_hits = 0
    end

    def get_row
      return @row
    end

    def get_col
      return @col
    end

    def get_number
      return @number
    end

    def set_pos(row,col)
      @row = row
      @col = col
    end

    def dead
      return @health < 1
    end

    def move(direction,valid_moves)
      size = valid_moves.size
      contained = valid_moves.include?(direction)
      if size > 0 && !contained
        return valid_moves[0]
      else
        return direction
      end
    end

    def attack
      return @strength + sum_weapons
    end

    def defend(received_attack)
      manage_hit(received_attack)
    end

    def receive_reward
      w_reward = Dice.weapons_reward
      s_reward = Dice.shields_reward
      for i in 1..w_reward
        wnew = new_weapon
        receive_weapon(wnew)
      end

      for i in 1..s_reward
        snew = new_shield
        receive_shield(snew)
      end

      extrahealth = Dice.health_reward
      @health += extrahealth
    end

    def to_s
      devolver =  "Nombre: #{@name}  Intelligence: #{@intelligence}  Strength: #{@strength} Health: #{@health} Position: [#{@row},#{@col}] consecutive hits: #{@consecutive_hits}  \n"
      devolver+= "Weapons: \n"
      if @weapons.length > 0
        for i in 0...@weapons.length
          devolver += @weapons[i]
        end
        devolver += "Shields: \n"
        for i in 0...@shields.length
          devolver += @shields[i]
        end
      end
      return devolver
    end

    private

    def receive_weapon(w)

      for i in 0...@weapons.length
        wi = @weapons[i]
        discard = wi.discard
        if discard
          @weapons.delete(wi)
        end
      end
      size = @weapons.length
      if size < @@MAX_WEAPONS
        @weapons.push(wi)
      end
    end

    def receive_shield(s)
      for i in 0...@shields.length
        si = @shields[i]
        discard = s.discard
        if discard
          @shields.delete(si)
        end
      end
      size = @shields.length
      if size < @@MAX_SHIELDS
        @shields.push(si)
      end
    end

    def new_weapon
      w= Weapon.new(Dice.weapons_power,Dice.uses_left)
      return w
    end

    def new_shield
      s= Shield.new(Dice.weapons_power,Dice.uses_left)
      return s
    end

    def sum_weapons
      total = 0.0
      for i in 0...@weapons.length
        total += @weapons[i]
      end
      return total
    end

    def sum_shields
      total = 0.0
      for i in 0...@shields.length
        total += @shields[i]
      end
      return total
    end

    def defensive_energy
      return @intelligence + sum_shields
    end

    def manage_hit(received_attack)
      defense = defensive_energy
      if defense < received_attack
        got_wounded
        inc_consecutive_hits
      else
        reset_hits
      end

      if (@consecutive_hits == @@HITS2LOSE) || dead
        lose = true
        reset_hits
      else
        lose = false
      end
      return lose
    end

    def reset_hits
      @consecutive_hits = 0
    end

    def got_wounded
      @health-=1
    end

    def inc_consecutive_hits
      @consecutive_hits += 1
    end

  end
end